src/main/scala/io/iohk/ethereum/network/PeerManagerActor.scala

```diff
@@ -99,37 +114,61 @@ class PeerManagerActor(
         log.debug("The known nodes list is empty")
       }
 
+    case PeerDiscoveryManager.RandomNodeInfo(node) =>
+      maybeConnectToRandomNode(connectedPeers, node)
+
     case PeerDiscoveryManager.DiscoveredNodesInfo(nodes) =>
-      val nodesToConnect = nodes
-        .filterNot { node =>
-          val socketAddress = node.tcpSocketAddress
-          val alreadyConnected =
-            connectedPeers.isConnectionHandled(socketAddress) || connectedPeers.hasHandshakedWith(node.id)
-          alreadyConnected || isBlacklisted(PeerAddress(socketAddress.getHostString))
-        } // not already connected to or blacklisted
-        .take(peerConfiguration.maxOutgoingPeers - connectedPeers.outgoingPeersCount)
-
-      NetworkMetrics.DiscoveredPeersSize.set(nodes.size)
-      NetworkMetrics.BlacklistedPeersSize.set(blacklistedPeers.size)
-      NetworkMetrics.PendingPeersSize.set(connectedPeers.pendingPeersCount)
-
-      log.info(
-        s"Discovered ${nodes.size} nodes, " +
-          s"Blacklisted ${blacklistedPeers.size} nodes, " +
-          s"handshaked to ${connectedPeers.handshakedPeersCount}/${peerConfiguration.maxOutgoingPeers + peerConfiguration.maxIncomingPeers}, " +
-          s"pending connection attempts ${connectedPeers.pendingPeersCount}. " +
-          s"Trying to connect to ${nodesToConnect.size} more nodes."
-      )
+      maybeConnectToDiscoveredNodes(connectedPeers, nodes)
+  }
 
-      if (nodesToConnect.nonEmpty) {
-        log.debug("Trying to connect to {} nodes", nodesToConnect.size)
-        nodesToConnect.foreach(n => self ! ConnectToPeer(n.toUri))
+  private def maybeConnectToRandomNode(connectedPeers: ConnectedPeers, node: Node): Unit = {
+    if (connectedPeers.outgoingConnectionDemand > 0) {
+      if (connectedPeers.canConnectTo(node)) {
+        log.debug(s"Trying to connect to random node at ${node.addr}")
+        self ! ConnectToPeer(node.toUri)
```

Comment:
Sorry I may be misunderstanding your question, I think we are doing exactly that:
```scala
   if (outgoingConnectionDemand > 0) {
      if (canConnectTo(node)) {
        // connect
      } else {
        // ask for another
        // Isn't this what you are asking? Demand is > 0 but we can't connect, so we ask for a replacement.
      }
    } else {
      // ignore because outgoingConnectionDemand <= 0
    }
```